criu/kerndat.c

```diff
@@ -1260,6 +1333,10 @@ int kerndat_init(void)
 		pr_err("kerndat_get_shmemdev failed when initializing kerndat.\n");
 		ret = -1;
 	}
+	if (!ret && kerndat_get_hugetlb_dev()) {
+		pr_err("kerndat_get_hugetlb_dev failed when initializing kerndat.\n");
+		ret = -1;
+	}
```

Comment:
This call seems redundant here. IMO, it would better to rework logic aroind re-detection and cache loading so that we won't need the repeated call to `kerndat_get_hugetlb_dev()`

Reply:
I think this call is needed at the first time we collect kerndat or when we fail to load the cache

Reply:
What I meant is that there is a call `kerndat_get_hugetlb_dev()` if we succeeded loading the cache, and then the call here when the cache is not available. We might leave only the call `kerndat_get_hugetlb_dev()` in the beginning of the function and adjust results checks accordingly.

Reply:
I have looked at bringing this call to the beginning of the function but in my opinion it adds more complexity but not so necessary. The reason is that there is another case when loading cache is that the cache is corrupted and we have to collect all the things again. Currently, `kerndat_get_hugetlb_dev()` collect directly into kdat, but if we move this before loading cache, we may need to change `kerndat_get_hugetlb_dev()` to load to a temporary buffer instead than copy to kdat later. Moreover, I change `kerndat_get_hugetlb_dev()` a little bit, so it return 0 on successful but no new hugetlb device's number is collect, 1 if successful and we collect new device's numbers and -1 on error. With this, I need to place this function after loading cache.